Validate `code` tags for class and member references

YuriSizov opened the pull request:
Inspired by changes in https://github.com/godotengine/godot/pull/82681 I have added checks for such cases where `[code]ClassName[/code]` should be replaced with a class reference, as well as similar checks for class members/methods/etc. As this is an uncertain check, it's marked as a warning. Which leads me to the second change...

This commit also, finally, adds means to manually disable warnings in `code` tags. These warnings can be false positive, and in fact there are about 30 cases where they certainly are. So far we've just ignored warnings on the CI and printed them out. But that's a lot of noise and it hides new warnings, and nobody is going to pay attention. So `code`, as well as `gdscript`, `csharp`, and `codeblock` can now have an optional `skip-lint` argument, which turns these warnings off. Use it wisely.

I've updated all problematic descriptions (see some of the notes below though), and updated all false positives to use the new argument. This allows us to enable warnings on CI, which is what I have done with this PR. I'm keeping them warnings and not errors to draw a semantic difference: errors are pretty much certain, whereas warnings are potential but need to be investigated.

-----

On the technical side of things, I reworked how we parse tags for RST a bit. I made sure there were no regressions in the generated RST files.

YuriSizov committed:
Validate `code` tags for class and member references

This commit also adds means to manually disable warnings
in `code` tags where it's a false positive with the new
`skip-lint` attribute.

Warnings are now enabled on CI to prevent future errors.

YuriSizov commented on doc/classes/Curve3D.xml:
```diff
@@ -138,7 +136,7 @@
 			<param index="2" name="apply_tilt" type="bool" default="false" />
 			<description>
-				Similar with [code]interpolate_baked()[/code]. The return value is [code]Transform3D[/code], with [code]origin[/code] as point position, [code]basis.x[/code] as sideway vector, [code]basis.y[/code] as up vector, [code]basis.z[/code] as forward vector. When the curve length is 0, there is no reasonable way to calculate the rotation, all vectors aligned with global space axes.
+				Returns a [Transform3D] with [code]origin[/code] as point position, [code]basis.x[/code] as sideway vector, [code]basis.y[/code] as up vector, [code]basis.z[/code] as forward vector. When the curve length is 0, there is no reasonable way to calculate the rotation, all vectors aligned with global space axes. See also [method sample_baked].
```
This method was renamed long ago, but this was missed. I should try to add checks for these as well.

YuriSizov replied:
Well, I tried, and caught 118 more potential issues. Some may be bad captures, as I cannot rule out a possibility of making mistakes in code or not accounting for something, but yeah... I won't be adding it to this PR just yet, can do it as a follow-up.

<details>
```

WARNING: @GDScript.xml: Found a string "@icon" that matches the @GDScript.@icon annotation in annotation "@icon" description.
WARNING: @GDScript.xml: Found a string "@icon" that matches the @GDScript.@icon annotation in annotation "@icon" description.
WARNING: @GDScript.xml: Found a string "@tool" that matches the @GDScript.@tool annotation in annotation "@tool" description.
WARNING: @GlobalScope.xml: Found a string "cubic_interpolate" that matches the @GlobalScope.cubic_interpolate method in method "cubic_interpolate_angle_in_time" description.
WARNING: @GlobalScope.xml: Found a string "Vector2.ZERO" that matches the Vector2.ZERO constant in method "type_convert" description.
WARNING: @GlobalScope.xml: Found a string "wrapf" that matches the @GlobalScope.wrapf method in method "wrapf" description.
WARNING: AStar3D.xml: Found a string "_compute_cost" that matches the AStar3D._compute_cost method in class "AStar3D" description.
WARNING: AStar3D.xml: Found a string "reserve_space" that matches the AStar3D.reserve_space method in method "get_point_capacity" description.
WARNING: AudioStreamPlayer.xml: Found a string "stream_paused" that matches the AudioStreamPlayer.stream_paused member in property "stream_paused" description.
WARNING: AudioStreamPlayer2D.xml: Found a string "stream_paused" that matches the AudioStreamPlayer2D.stream_paused member in property "stream_paused" description.
WARNING: BaseMaterial3D.xml: Found a string "distance_fade_max_distance" that matches the BaseMaterial3D.distance_fade_max_distance member in property "distance_fade_max_distance" description.
WARNING: BaseMaterial3D.xml: Found a string "distance_fade_min_distance" that matches the BaseMaterial3D.distance_fade_min_distance member in property "distance_fade_max_distance" description.
WARNING: BaseMaterial3D.xml: Found a string "distance_fade_max_distance" that matches the BaseMaterial3D.distance_fade_max_distance member in property "distance_fade_max_distance" description.
WARNING: BaseMaterial3D.xml: Found a string "distance_fade_min_distance" that matches the BaseMaterial3D.distance_fade_min_distance member in property "distance_fade_max_distance" description.
WARNING: BaseMaterial3D.xml: Found a string "distance_fade_min_distance" that matches the BaseMaterial3D.distance_fade_min_distance member in property "distance_fade_min_distance" description.
WARNING: BaseMaterial3D.xml: Found a string "distance_fade_max_distance" that matches the BaseMaterial3D.distance_fade_max_distance member in property "distance_fade_min_distance" description.
WARNING: BaseMaterial3D.xml: Found a string "distance_fade_max_distance" that matches the BaseMaterial3D.distance_fade_max_distance member in property "distance_fade_min_distance" description.
WARNING: BaseMaterial3D.xml: Found a string "distance_fade_min_distance" that matches the BaseMaterial3D.distance_fade_min_distance member in property "distance_fade_min_distance" description.
WARNING: Basis.xml: Found a string "Basis" that matches one of the known classes in constant "IDENTITY" description.
WARNING: Basis.xml: Found a string "Basis" that matches the Basis.Basis constructor in constant "IDENTITY" description.
WARNING: Button.xml: Found a string "focus" that matches the Button.focus theme item in theme item "focus" description.
WARNING: Camera3D.xml: Found a string "fov" that matches the Camera3D.fov member in property "fov" description.
WARNING: CharacterBody2D.xml: Found a string "Vector2.UP" that matches the Vector2.UP constant in property "up_direction" description.
WARNING: CharacterBody2D.xml: Found a string "Vector2.UP" that matches the Vector2.UP constant in method "get_floor_angle" description.
WARNING: CharacterBody3D.xml: Found a string "Vector3.UP" that matches the Vector3.UP constant in property "up_direction" description.
WARNING: CharacterBody3D.xml: Found a string "Vector3.UP" that matches the Vector3.UP constant in method "get_floor_angle" description.
WARNING: CodeEdit.xml: Found a string "indent_use_spaces" that matches the CodeEdit.indent_use_spaces member in property "indent_size" description.
WARNING: CollisionObject2D.xml: Found a string "input_event" that matches the CollisionObject2D.input_event signal in method "_input_event" description.
WARNING: CollisionPolygon3D.xml: Found a string "polygon" that matches the CollisionPolygon3D.polygon member in property "polygon" description.
WARNING: DampedSpringJoint2D.xml: Found a string "damping" that matches the DampedSpringJoint2D.damping member in property "damping" description.
WARNING: DisplayServer.xml: Found a string "FEATURE_NATIVE_DIALOG" that matches the DisplayServer.FEATURE_NATIVE_DIALOG enum value in method "file_dialog_show" description.
WARNING: EditorExportPlatformMacOS.xml: Found a string "privacy/address_book_usage_description" that matches the EditorExportPlatformMacOS.privacy/address_book_usage_description member in property "codesign/entitlements/address_book" description.
WARNING: EditorExportPlatformMacOS.xml: Found a string "privacy/microphone_usage_description" that matches the EditorExportPlatformMacOS.privacy/microphone_usage_description member in property "codesign/entitlements/audio_input" description.
WARNING: EditorExportPlatformMacOS.xml: Found a string "privacy/calendar_usage_description" that matches the EditorExportPlatformMacOS.privacy/calendar_usage_description member in property "codesign/entitlements/calendars" description.
WARNING: EditorExportPlatformMacOS.xml: Found a string "privacy/camera_usage_description" that matches the EditorExportPlatformMacOS.privacy/camera_usage_description member in property "codesign/entitlements/camera" description.
WARNING: EditorExportPlatformMacOS.xml: Found a string "privacy/location_usage_description" that matches the EditorExportPlatformMacOS.privacy/location_usage_description member in property "codesign/entitlements/location" description.
WARNING: EditorExportPlatformMacOS.xml: Found a string "privacy/photos_library_usage_description" that matches the EditorExportPlatformMacOS.privacy/photos_library_usage_description member in property "codesign/entitlements/photos_library" description.
WARNING: EditorExportPlugin.xml: Found a string "add_ios_framework" that matches the EditorExportPlugin.add_ios_framework method in method "add_ios_embedded_framework" description.
WARNING: Environment.xml: Found a string "adjustment_enabled" that matches the Environment.adjustment_enabled member in property "adjustment_brightness" description.
WARNING: Environment.xml: Found a string "adjustment_enabled" that matches the Environment.adjustment_enabled member in property "adjustment_color_correction" description.
WARNING: Environment.xml: Found a string "adjustment_enabled" that matches the Environment.adjustment_enabled member in property "adjustment_contrast" description.
WARNING: Environment.xml: Found a string "adjustment_enabled" that matches the Environment.adjustment_enabled member in property "adjustment_saturation" description.
WARNING: GPUParticles3D.xml: Found a string "process_material" that matches the GPUParticles3D.process_material member in class "GPUParticles3D" description.
WARNING: GPUParticles3D.xml: Found a string "amount" that matches the GPUParticles3D.amount member in property "one_shot" description.
WARNING: HTTPRequest.xml: Found a string "request_completed" that matches the HTTPRequest.request_completed signal in property "accept_gzip" description.
WARNING: HTTPRequest.xml: Found a string "accept_gzip" that matches the HTTPRequest.accept_gzip member in property "accept_gzip" description.
WARNING: HTTPRequest.xml: Found a string "request_completed" that matches the HTTPRequest.request_completed signal in property "accept_gzip" description.
WARNING: Image.xml: Found a string "data" that matches the Image.data member in method "get_mipmap_offset" description.
WARNING: JavaScriptBridge.xml: Found a string "eval" that matches the JavaScriptBridge.eval method in method "eval" description.
WARNING: Joint2D.xml: Found a string "bias" that matches the Joint2D.bias member in property "bias" description.
WARNING: Joint2D.xml: Found a string "bias" that matches the Joint2D.bias member in property "bias" description.
WARNING: Label.xml: Found a string "lines_skipped" that matches the Label.lines_skipped member in property "lines_skipped" description.
WARNING: Line2D.xml: Found a string "texture_mode" that matches the Line2D.texture_mode member in property "texture" description.
WARNING: LinkButton.xml: Found a string "focus" that matches the LinkButton.focus theme item in theme item "focus" description.
WARNING: MarginContainer.xml: Found a string "margin_bottom" that matches the MarginContainer.margin_bottom theme item in theme item "margin_bottom" description.
WARNING: MarginContainer.xml: Found a string "margin_left" that matches the MarginContainer.margin_left theme item in theme item "margin_left" description.
WARNING: MarginContainer.xml: Found a string "margin_right" that matches the MarginContainer.margin_right theme item in theme item "margin_right" description.
WARNING: MarginContainer.xml: Found a string "margin_top" that matches the MarginContainer.margin_top theme item in theme item "margin_top" description.
WARNING: MenuButton.xml: Found a string "switch_on_hover" that matches the MenuButton.switch_on_hover member in property "switch_on_hover" description.
WARNING: NavigationPathQueryParameters2D.xml: Found a string "map" that matches the NavigationPathQueryParameters2D.map member in property "map" description.
WARNING: NavigationPathQueryParameters3D.xml: Found a string "map" that matches the NavigationPathQueryParameters3D.map member in property "map" description.
WARNING: NavigationRegion2D.xml: Found a string "enter_cost" that matches the NavigationRegion2D.enter_cost member in property "enter_cost" description.
WARNING: NavigationRegion2D.xml: Found a string "travel_cost" that matches the NavigationRegion2D.travel_cost member in property "travel_cost" description.
WARNING: NavigationRegion3D.xml: Found a string "enter_cost" that matches the NavigationRegion3D.enter_cost member in property "enter_cost" description.
WARNING: NavigationRegion3D.xml: Found a string "travel_cost" that matches the NavigationRegion3D.travel_cost member in property "travel_cost" description.
WARNING: Node.xml: Found a string "_ready" that matches the Node._ready method in method "_ready" description.
WARNING: Node.xml: Found a string "_ready" that matches the Node._ready method in method "request_ready" description.
WARNING: Node.xml: Found a string "_ready" that matches the Node._ready method in method "request_ready" description.
WARNING: Node.xml: Found a string "_ready" that matches the Node._ready method in method "request_ready" description.
WARNING: Node.xml: Found a string "_ready" that matches the Node._ready method in method "request_ready" description.
WARNING: Object.xml: Found a string "Object.connect" that matches the Object.connect method in method "connect" description.
WARNING: Object.xml: Found a string "Signal.connect" that matches the Signal.connect method in method "connect" description.
WARNING: PackedByteArray.xml: Found a string "decompress" that matches the PackedByteArray.decompress method in method "decompress_dynamic" description.
WARNING: PackedByteArray.xml: Found a string "decompress" that matches the PackedByteArray.decompress method in method "decompress_dynamic" description.
WARNING: PackedScene.xml: Found a string "pack" that matches the PackedScene.pack method in class "PackedScene" description.
WARNING: PacketPeer.xml: Found a string "encode_buffer_max_size" that matches the PacketPeer.encode_buffer_max_size member in property "encode_buffer_max_size" description.
WARNING: ParticleProcessMaterial.xml: Found a string "emission_shape" that matches the ParticleProcessMaterial.emission_shape member in property "emission_box_extents" description.
WARNING: ParticleProcessMaterial.xml: Found a string "emission_shape" that matches the ParticleProcessMaterial.emission_shape member in property "emission_point_count" description.
WARNING: ParticleProcessMaterial.xml: Found a string "emission_shape" that matches the ParticleProcessMaterial.emission_shape member in property "emission_sphere_radius" description.
WARNING: PhysicsServer2D.xml: Found a string "RID" that matches one of the known classes in method "area_get_space" description.
WARNING: PhysicsServer2D.xml: Found a string "RID" that matches one of the known classes in method "body_get_space" description.
WARNING: PhysicsServer2D.xml: Found a string "RID" that matches one of the known classes in method "joint_make_pin" description.
WARNING: PhysicsServer3D.xml: Found a string "body_set_force_integration_callback" that matches the PhysicsServer3D.body_set_force_integration_callback method in method "body_set_force_integration_callback" description.
WARNING: Plane.xml: Found a string "d" that matches the Plane.d member in property "d" description.
WARNING: Plane.xml: Found a string "d" that matches the Plane.d member in property "normal" description.
WARNING: Polygon2D.xml: Found a string "texture" that matches the Polygon2D.texture member in property "color" description.
WARNING: Polygon2D.xml: Found a string "vertex_colors" that matches the Polygon2D.vertex_colors member in property "color" description.
WARNING: Polygon2D.xml: Found a string "uv" that matches the Polygon2D.uv member in property "texture" description.
WARNING: Polygon2D.xml: Found a string "texture" that matches the Polygon2D.texture member in property "texture_offset" description.
WARNING: Polygon2D.xml: Found a string "uv" that matches the Polygon2D.uv member in property "texture_scale" description.
WARNING: Polygon2D.xml: Found a string "texture" that matches the Polygon2D.texture member in property "texture_scale" description.
WARNING: Polygon2D.xml: Found a string "uv" that matches the Polygon2D.uv member in property "uv" description.
WARNING: Polygon2D.xml: Found a string "color" that matches the Polygon2D.color member in property "vertex_colors" description.
WARNING: ProjectSettings.xml: Found a string "xr/openxr/foveation_level" that matches the ProjectSettings.xr/openxr/foveation_level member in property "xr/openxr/foveation_dynamic" description.
WARNING: Quaternion.xml: Found a string "spherical_cubic_interpolate" that matches the Quaternion.spherical_cubic_interpolate method in method "spherical_cubic_interpolate_in_time" description.
WARNING: RichTextLabel.xml: Found a string "text" that matches the RichTextLabel.text member in property "text" description.
WARNING: RichTextLabel.xml: Found a string "focus" that matches the RichTextLabel.focus theme item in theme item "focus" description.
WARNING: RigidBody3D.xml: Found a string "Vector3.ZERO" that matches the Vector3.ZERO constant in property "inertia" description.
WARNING: Slider.xml: Found a string "grabber_area" that matches the Slider.grabber_area theme item in theme item "slider" description.
WARNING: SpinBox.xml: Found a string "custom_arrow_step" that matches the SpinBox.custom_arrow_step member in property "custom_arrow_step" description.
WARNING: SpinBox.xml: Found a string "prefix" that matches the SpinBox.prefix member in property "prefix" description.
WARNING: SpinBox.xml: Found a string "suffix" that matches the SpinBox.suffix member in property "suffix" description.
WARNING: TextEdit.xml: Found a string "focus" that matches the TextEdit.focus theme item in theme item "focus" description.
WARNING: TextureProgressBar.xml: Found a string "texture_over" that matches the TextureProgressBar.texture_over member in property "tint_over" description.
WARNING: TextureProgressBar.xml: Found a string "texture_progress" that matches the TextureProgressBar.texture_progress member in property "tint_progress" description.
WARNING: TextureProgressBar.xml: Found a string "texture_under" that matches the TextureProgressBar.texture_under member in property "tint_under" description.
WARNING: TileSet.xml: Found a string "add_source" that matches the TileSet.add_source method in method "get_next_source_id" description.
WARNING: TileSetAtlasSource.xml: Found a string "margins" that matches the TileSetAtlasSource.margins member in method "get_atlas_grid_size" description.
WARNING: TileSetAtlasSource.xml: Found a string "texture_region_size" that matches the TileSetAtlasSource.texture_region_size member in method "get_atlas_grid_size" description.
WARNING: Transform2D.xml: Found a string "is_equal_approx" that matches the Transform2D.is_equal_approx method in method "is_equal_approx" description.
WARNING: Transform3D.xml: Found a string "is_equal_approx" that matches the Transform3D.is_equal_approx method in method "is_equal_approx" description.
WARNING: Tree.xml: Found a string "item_selected" that matches the Tree.item_selected signal in signal "multi_selected" description.
WARNING: Tree.xml: Found a string "select_mode" that matches the Tree.select_mode member in signal "multi_selected" description.
WARNING: Vector2.xml: Found a string "cubic_interpolate" that matches the Vector2.cubic_interpolate method in method "cubic_interpolate_in_time" description.
WARNING: Vector3.xml: Found a string "cubic_interpolate" that matches the Vector3.cubic_interpolate method in method "cubic_interpolate_in_time" description.
WARNING: Vector4.xml: Found a string "cubic_interpolate" that matches the Vector4.cubic_interpolate method in method "cubic_interpolate_in_time" description.
WARNING: VisualShaderNodeSwitch.xml: Found a string "op_type" that matches the VisualShaderNodeSwitch.op_type member in class "VisualShaderNodeSwitch" description.
WARNING: WorldBoundaryShape2D.xml: Found a string "Vector2.UP" that matches the Vector2.UP constant in property "normal" description.

Generating the index file...

118 warnings were found in the class reference XML. Please check the messages above.
```
</details>

YuriSizov replied:
Well, I just fixed all of them anyway.

YuriSizov commented on doc/classes/FontVariation.xml:
```diff
@@ -70,7 +70,7 @@
 		<member name="variation_opentype" type="Dictionary" setter="set_variation_opentype" getter="get_variation_opentype" default="{}">
 			Font OpenType variation coordinates. More info: [url=https://docs.microsoft.com/en-us/typography/opentype/spec/dvaraxisreg]OpenType variation tags[/url].
-			[b]Note:[/b] This [Dictionary] uses OpenType tags as keys. Variation axes can be identified both by tags([code]int[/code]) and names ([code]string[/code]). Some axes might be accessible by multiple names. For example, [code]wght[/code] refers to the same axis as [code]weight[/code]. Tags on the other hand are unique. To convert between names and tags, use [method TextServer.name_to_tag] and [m ...
+			[b]Note:[/b] This [Dictionary] uses OpenType tags as keys. Variation axes can be identified both by tags ([code safe]int[/code]) and names ([code safe]string[/code]). Some axes might be accessible by multiple names. For example, [code]wght[/code] refers to the same axis as [code]weight[/code]. Tags on the other hand are unique. To convert between names and tags, use [method TextServer.name_to_ ...
```
Not entirely sure what these are referring to here, maybe it should be a reference?

bruvzg replied:
These probably should be base types, `[int]` and `[string]`. It's referring to the type of variable that can be used as a Dictionary key. e.g., readable name - "weight" or short name - "wght" or integer tag - 0x77678674 (ASCII for "wght" interpreter as int).

YuriSizov commented on doc/classes/ImageTextureLayered.xml:
```diff
@@ -23,7 +23,7 @@
 			<description>
 				Replaces the existing [Image] data at the given [param layer] with this new image.
-				The given [Image] must have the same width, height, image format and mipmapping setting (a [code]bool[/code] value) as the rest of the referenced images.
+				The given [Image] must have the same width, height, image format, and mipmapping flag as the rest of the referenced images.
```
Phrasing felt a bit weird, so I just reworded it.

YuriSizov commented on doc/classes/DisplayServer.xml:
```diff
@@ -1055,7 +1055,7 @@
 				- [code]name[/code] is voice name.
 				- [code]id[/code] is voice identifier.
-				- [code]language[/code] is language code in [code]lang_Variant[/code] format. [code]lang[/code] part is a 2 or 3-letter code based on the ISO-639 standard, in lowercase. And [code]Variant[/code] part is an engine dependent string describing country, region or/and dialect.
+				- [code]language[/code] is language code in [code]lang_Variant[/code] format. The [code]lang[/code] part is a 2 or 3-letter code based on the ISO-639 standard, in lowercase. The [code safe]Variant[/code] part is an engine dependent string describing country, region or/and dialect.
```
This seems to imitate the casing of the value itself, which I don't think is very clear in this case. But it matches a similar example in OS.xml, so ultimately I left this one as is.

akien-mga commented on doc/tools/make_rst.py:
```diff
@@ -1721,17 +1778,19 @@ def format_text_block(
                 inside_code = True
-                inside_code_tag = cmd
+                inside_code_tag = tag_state.name
+                ignore_code_warnings = "safe" in tag_state.arguments
```
As discussed on RC, I would suggest using `nolint`, or another option along those lines. Could also be `nowarn` or `skip-lint` or something like this.

`safe` seems a bit weird to me in the context of `code` tags.

YuriSizov replied:
I like `skip-lint`, will update to it.

MewPurPur commented on doc/classes/NavigationMeshSourceGeometryData3D.xml:
```diff
@@ -14,23 +14,23 @@
 			<param index="1" name="xform" type="Transform3D" />
 			<description>
-				Adds an array of vertex positions to the geometry data for navigation mesh baking to form triangulated faces. For each face the array must have three vertex positions in clockwise winding order. Since [NavigationMesh] resource have no transform all vertex positions need to be offset by the node's transform using the [code]xform[/code] parameter.
+				Adds an array of vertex positions to the geometry data for navigation mesh baking to form triangulated faces. For each face the array must have three vertex positions in clockwise winding order. Since [NavigationMesh] resource have no transform all vertex positions need to be offset by the node's transform using [param xform].
```
```suggestion
				Adds an array of vertex positions to the geometry data for navigation mesh baking to form triangulated faces. For each face the array must have three vertex positions in clockwise winding order. Since [NavigationMesh] resources have no transform, all vertex positions need to be offset by the node's transform using [param xform].
```

Likewise below.

bruvzg commented on doc/classes/FileAccess.xml:
```diff
@@ -341,7 +341,7 @@
 			<param index="1" name="hidden" type="bool" />
 			<description>
-				Sets file [code]hidden[/code] attribute.
+				Sets file [code safe]hidden[/code] attribute.
```
Maybe
```suggestion
				Sets file [b]hidden[/b] attribute.
```

It's not a code, just used to highlight it.

YuriSizov commented:
What is "scope creep", Alex. I've added another set of checks so we can catch not only `[code]ClassName[/code]`, but also `[code]some_member[/code]` and even `[code]some_method()[/code]`, which are supposed to be references. There was around  120 matches, a decent chunk of which was self-references (so a minor crime). But also there were a lot of genuine issues, some even worth rewording descriptions. I didn't go out of my way to fix the docs, just made changes necessary to address the warnings caught.

I also realized that the new `skip-lint` directive needs to be supported by `EditorHelp`, which is now fixed. Another addition to the built-in help is some nominal support for operator and constructor references, something that I was planning to check [long time ago](https://github.com/godotengine/godot/pull/55054#discussion_r751186812). `Basis` actually has a use case for constructors, so I decided to fix it alongside everything as well.

Other suggestions have also been implemented, and this should be good for a review now.

raulsntos commented:
> I also realized that the new `skip-lint` directive needs to be supported by `EditorHelp`

Also needs to be changed in the C# bindings generator to avoid breaking documentation generation, probably something like this may be enough (haven't tested it):

```diff
diff --git a/modules/mono/editor/bindings_generator.cpp b/modules/mono/editor/bindings_generator.cpp
index ccfcf2a87c8..8b97dc8a98c 100644
--- a/modules/mono/editor/bindings_generator.cpp
+++ b/modules/mono/editor/bindings_generator.cpp
@@ -401,29 +401,29 @@ String BindingsGenerator::bbcode_to_xml(const String &p_bbcode, const TypeInterf
 
 			pos = brk_end + 1;
 			tag_stack.push_front(tag);
-		} else if (tag == "code") {
+		} else if (tag == "code" || tag.begins_with("code ")) {
 			xml_output.append("<c>");
 
 			code_tag = true;
 			pos = brk_end + 1;
-			tag_stack.push_front(tag);
-		} else if (tag == "codeblock") {
+			tag_stack.push_front("code");
+		} else if (tag == "codeblock" || tag.begins_with("codeblock ")) {
 			xml_output.append("<code>");
 
 			code_tag = true;
 			pos = brk_end + 1;
-			tag_stack.push_front(tag);
+			tag_stack.push_front("codeblock");
 		} else if (tag == "codeblocks") {
 			line_del = true;
 			pos = brk_end + 1;
 			tag_stack.push_front(tag);
-		} else if (tag == "csharp") {
+		} else if (tag == "csharp" || tag.begins_with("csharp ")) {
 			xml_output.append("<code>");
 
 			line_del = false;
 			code_tag = true;
 			pos = brk_end + 1;
-			tag_stack.push_front(tag);
+			tag_stack.push_front("csharp");
 		} else if (tag == "kbd") {
 			// keyboard combinations are not supported in xml comments
 			pos = brk_end + 1;

```

YuriSizov commented:
@raulsntos How can I test it? Just generate glue and check relevant XML documentation comments?

raulsntos commented:
Yeah, if there's malformed XML documentation there should also be warnings when building the C# projects but it'll likely be impossible to see them among the many warnings that we currently have.

akien-mga commented on doc/classes/CollisionObject2D.xml:
```diff
@@ -16,7 +16,7 @@
 			<param index="2" name="shape_idx" type="int" />
 			<description>
-				Accepts unhandled [InputEvent]s. [param shape_idx] is the child index of the clicked [Shape2D]. Connect to the [code]input_event[/code] signal to easily pick up these events.
+				Accepts unhandled [InputEvent]s. [param shape_idx] is the child index of the clicked [Shape2D]. Connect to [signal input_event] to easily pick up these events.
```
Random thought, but since this name might be ambiguous, I understand why one might want to make "signal" explicit in the text so it's not confused to methods like `_input_event` in this same class.

I'm not advocating to change it back, but it makes me think that we should probably aim to have a tooltip on all those hyperlinks providing context on the type of API, and maybe its short description.

This would prevent having to click a link to know what it's referring to exactly, before going back in history where one was.

Totally for a separate PR / proposal, and after Yeldham's refactor is merged.

YuriSizov replied:
> I'm not advocating to change it back, but it makes me think that we should probably aim to have a tooltip on all those hyperlinks providing context on the type of API, and maybe its short description.

I was thinking an icon, but we should also be able to add tooltips now, yeah.

akien-mga commented on doc/classes/DisplayServer.xml:
```diff
@@ -1055,7 +1055,7 @@
 				- [code]name[/code] is voice name.
 				- [code]id[/code] is voice identifier.
-				- [code]language[/code] is language code in [code]lang_Variant[/code] format. [code]lang[/code] part is a 2 or 3-letter code based on the ISO-639 standard, in lowercase. And [code]Variant[/code] part is an engine dependent string describing country, region or/and dialect.
+				- [code]language[/code] is language code in [code]lang_Variant[/code] format. The [code]lang[/code] part is a 2 or 3-letter code based on the ISO-639 standard, in lowercase. The [code skip-lint]Variant[/code] part is an engine dependent string describing country, region or/and dialect.
```
```suggestion
				- [code]language[/code] is language code in [code]lang_Variant[/code] format. The [code]lang[/code] part is a 2 or 3-letter code based on the ISO-639 standard, in lowercase. The [code skip-lint]Variant[/code] part is an engine-dependent string describing country, region or/and dialect.
```

akien-mga commented on doc/classes/MarginContainer.xml:
```diff
@@ -30,16 +30,16 @@
 	<theme_items>
 		<theme_item name="margin_bottom" data_type="constant" type="int" default="0">
-			All direct children of [MarginContainer] will have a bottom margin of [code]margin_bottom[/code] pixels.
+			Offsets towards the inside direct children of the container by this amount of pixels from the bottom.
```
I agree with rephrasing, but I think the new one is still hard to parse.
It took me multiple attempts to break down in individual semantic elements and understand the flow of that sentence. "Offsets towards the inside" used as a verb especially makes it complex.

YuriSizov replied:
Yes, that one is hard to get right. I could use "insets" perhaps, instead of the whole "offsets towards the inside", though I consider the "inset" word to be a bit less clear to non-English speakers for no particular reason.

akien-mga commented on doc/classes/Object.xml:
```diff
@@ -487,7 +487,7 @@
 				[/csharp]
 				[/codeblocks]
-				[b][code]Object.connect()[/code] or [code]Signal.connect()[/code]?[/b]
+				[b][code skip-lint]Object.connect()[/code] or [code skip-lint]Signal.connect()[/code]?[/b]
```
Could possibly be:
```suggestion
				[b][method Object.connect] or [method Signal.connect]?[/b]
```
But I guess the former will only show as [connect]() since we're in Object?

YuriSizov replied:
Perhaps, but I just didn't want to break the original formatting since here it seems to be done on purpose for the title.

akien-mga commented on doc/classes/Polygon2D.xml:
```diff
@@ -100,22 +100,22 @@
 		</member>
 		<member name="texture_offset" type="Vector2" setter="set_texture_offset" getter="get_texture_offset" default="Vector2(0, 0)">
-			Amount to offset the polygon's [code]texture[/code]. If [code](0, 0)[/code] the texture's origin (its top-left corner) will be placed at the polygon's [code]position[/code].
+			Amount to offset the polygon's [member texture]. If set to [code]Vector2(0, 0)[/code] the texture's origin (its top-left corner) will be placed at the polygon's position.
```
```suggestion
			Amount to offset the polygon's [member texture]. If set to [code]Vector2(0, 0)[/code], the texture's origin (its top-left corner) will be placed at the polygon's position.
```

YuriSizov replied:
Oh, by the way, I forgot to mention this myself, but I have no idea what "polygon's position" is referring to. That may be something worth clarifying by someone who understands how this works.

akien-mga commented on editor/editor_help.cpp:
```diff
@@ -410,8 +416,10 @@ String EditorHelp::_fix_constant(const String &p_constant) const {
-	method_line[p_method.name] = class_desc->get_paragraph_count() - 2; // Gets overridden if description
+void EditorHelp::_add_method(const DocData::MethodDoc &p_method, bool p_overview, bool p_override) {
+	if (p_override) {
+		method_line[p_method.name] = class_desc->get_paragraph_count() - 2; // Gets overridden if description
```
```suggestion
		method_line[p_method.name] = class_desc->get_paragraph_count() - 2; // Gets overridden if description.
```

akien-mga reviewed (approved):
Looks great! Read through all changes, seems good to go.

YuriSizov commented:
Applied suggestions and added necessary changes to the C# bindings generator. Not sure how to improve "Offsets towards the inside", so left it as is for now.

raulsntos reviewed (approved):
The changes to the C# bindings generator look good to me.

I noticed a typo in the `Light3D` documentation while testing this. The `BAKE_DISABLED` constant is not closing a member tag. Not really related to this PR though, but here's a diff anyway:

```diff
diff --git a/doc/classes/Light3D.xml b/doc/classes/Light3D.xml
index ebb36ba92be..df3d802e16e 100644
--- a/doc/classes/Light3D.xml
+++ b/doc/classes/Light3D.xml
@@ -199,7 +199,7 @@
 		</constant>
 		<constant name="BAKE_DISABLED" value="0" enum="BakeMode">
 			Light is ignored when baking. This is the fastest mode, but the light will be taken into account when baking global illumination. This mode should generally be used for dynamic lights that change quickly, as the effect of global illumination is less noticeable on those lights.
-			[b]Note:[/b] Hiding a light does [i]not[/i] affect baking [LightmapGI]. Hiding a light will still affect baking [VoxelGI] and SDFGI (see [member Environment.sdfgi_enabled).
+			[b]Note:[/b] Hiding a light does [i]not[/i] affect baking [LightmapGI]. Hiding a light will still affect baking [VoxelGI] and SDFGI (see [member Environment.sdfgi_enabled]).
 		</constant>
 		<constant name="BAKE_STATIC" value="1" enum="BakeMode">
 			Light is taken into account in static baking ([VoxelGI], [LightmapGI], SDFGI ([member Environment.sdfgi_enabled])). The light can be moved around or modified, but its global illumination will not update in real-time. This is suitable for subtle changes (such as flickering torches), but generally not large changes such as toggling a light on and off.

```

akien-mga commented:
> I noticed a typo in the `Light3D` documentation while testing this. The `BAKE_DISABLED` constant is not closing a member tag.

This would be worth adding to the validation checks in `make_rst.py`, but that can be a follow-up PR.

akien-mga commented:
Thanks!
